### ISSUE TITLE
Handling load conflicts in a SCIMetacelloLoadSpec (and other Metacello Scripting API functions)

### ISSUE THREAD

**Issue by jbrichau** [hpi-swa/smalltalkCI#333]
I'm moving a project to Travis-CI where the Metacello Scripting API is required to handle load conflicts in an `onConflict:` handler. The project's baseline references many other project configurations and we need to handle several load conflicts (e.g. different projects reference different versions of Seaside) in order to load the code into a Pharo image.

We currently do that using a custom Metacello load script implemented on the class side of our BaselineOf class. This custom script also sets a `repositoryOverrides:` because we replicate all dependent packages to a folder in the same git repository. This makes our load independent from availability or changes in outside repositories.

Is there a way to do this in Smalltalk-CI ? I have been experimenting with the `#preLoading` script, which allows me to load the project but it seems I need to add a `#loading` declaration as well or it will not work. I think I can make a stub load group to make this work but it feels like a workaround, while I would actually want to use additional functions of the Metacello API.

If this is not available, I'm happy to take pointers to how this would be possible to implement and give a PR a go. We need to move this project to Travis-CI, so definitely interested to extend Smalltalk-CI if possible.

**Comment by fniephaus** [comment]
Hi @jbrichau,
Could you please provide such a Metacello load script for you project?
The `#preLoading` option is available since last week, so maybe it's a bug that `#loading` is required. I believe it shouldn't be required.
How do your users install your project? Do you make them handle the conflicts themselves? Or could you avoid any conflicts in a `preLoad` script in your baseline?

**Comment by jbrichau** [comment]
Hey @fniephaus,

The project is a private project (Yesplan), so I cannot share the entire config out-of-the-box. 
Here is the load script (which is a class method of our BaselineOf). Do not mind the exact filepath in there, I am experimenting to make this work :):

```Smalltalk
loadDevelopment
	"self loadDevelopment"
	| repository |
	repository := MCFileTreeRepository new directory: (FileSystem workingDirectory / 'git_cache' / 'repository').
	Metacello new
   		baseline: 'NextPlan';
   		repository: repository description;
		repositoryOverrides: (self otherRepositories add: repository description; yourself);
		onConflict: [:ex | 
    				(ex newProjectRegistration projectName = ex existingProjectRegistration projectName and:[ex newProjectRegistration version = ex existingProjectRegistration version]) 
					ifTrue: [ ex allow]
    				ifFalse:[ 
						(self isAllowedProject: ex newProjectRegistration projectName version: ex newProjectRegistration version in: #development)
							ifTrue:[ ex allow ]
							ifFalse: [ 
								Transcript show: 'Disallow ',ex newProjectRegistration projectName,' ',ex newProjectRegistration version.
								ex disallow ]]
					];
  		 load: 'NextPlan-Dev'
```

**Comment by fniephaus** [comment]
I don't really see how we could extend the config spec in such a way that a use case like this could be covered (with relatively low effort).
Anyhow, now that custom scripts are supported during pre/post loading/testing phases, maybe we could allow them in `#loading` as well. The problem is, that smalltalkCI is then unable to automatically determine with tests to run.
What happens if you only use `#preLoading` and define the tests to run manually?

**Comment by dalehenrich** [comment]
I don't have the time to look into this at the moment and my computer is being upgraded shortly, so I don't know exactly when it will be back ... but I don't think the preload approach is necessarily the right answer --- I need to study the preload stuff a bit more  --- since what @jbrichau is doing is a valid/long-standing part of the Metacello load api, it perhaps _should_ be part of the Metacello spec ... but I need to be able to spend more time studying ...

Out of curiosity, @jbrichau, I assume that you are using SmalltalkCI locally and not part of a Travis-CI run? You mention that you are using repositoryOverrides: and I am curious how you are doing that for travis --- if you are doing it for automated local builds then I understand the need ...

**Comment by jbrichau** [comment]
@fniephaus I'm going to try that asap (was a bit busy today) and let you know. I think it should be possible to extend SmalltalkCI to use a custom script for loading. In that case, it's acceptable that we should specify the tests manually.

@dalehenrich we want to use Travis-CI, though I have been testing locally most of the time. Indeed, another part of the puzzle is to know where the repository location is to specify it in a `repositoryOverrides:`. For now, I have been playing with the `git_cache` location I noticed exists on local builds. What I want to achieve is that we replicate projects we depend on and that are not on github into our own repository in a separate directory (e.g. `'filetree://<repo_location>/externals/'`). This is what we have been doing in the past as well to make sure we have a fully contained checkout on disk that does not require external repos to exist before we load the code into an image.

**Comment by dalehenrich** [comment]
@jbrichau .. ah that makes a lot of sense ... I guess there are still projects that haven't moved to GitHub:)? ... For my part I should be able to eliminate Gofer and Sport from list of Monticello-repo-based projects ... Are you using the XML parser/write project? It's on the radar, but I'm hoping that the pharo community moves it to GitHub first:)

**Comment by fniephaus** [comment]
Side note: the `git_cache` symlink will soon be removed (https://github.com/hpi-swa/smalltalkCI/commit/20987586d396f126830f934d134e492be4f24543 is already on `dev`), because it is no longer used by smalltalkCI. If you your project relies on other resources, you can copy them into the build directory (`SMALLTALK_CI_BUILD`) before smalltalkCI is executed.

**Comment by seandenigris** [comment]
@fniephaus @jbrichau Is it now possible to pass a custom script to `loading:`?

I ran into a similar problem re conflicts and asked on the Metacello ML:

> I have a GH fork of Magritte3 with many enhancements and fixes, on which I depend for many personal projects. I run into trouble when the "official" Magritte is loaded (by e.g. Glamour IIRC). In a manual loading script, I came up with the following:
```smalltalk
	onConflict: [ :ex | 
		ex existingProjectRegistration projectName = 'Magritte3' ifTrue: [
			ex newProjectRegistration baselineProjectSpec "The version we want is the one loaded via baseline, not the one via Configuration"
				ifNil: [ ex disallow ] 
				ifNotNil: [ ex allow ] ] ];
```
> Is the approach above okay? Is there a better way? And most importantly, how would this sort of information be passed to smalltalkCI? I don't see a hook for conflict resolution…

And @dalehenrich's answer was:

> Your approach is a good one ... and yes it seems that this a pretty good case for wanting to add an onConflict: block to the SCIMetacelloLoadSpec, so its' worth opening an issue against smalltalkCI ... I have already been pretty much hammered schedule wise but the next week or so will be even more frenetic for me ... after that though I will have the time to slip in some SmalltalkCI work, if Fabio hasn't already done something:)

**Comment by seandenigris** [comment]
I'm not sure why this worked, but locking Magritte3 even *before* Magritte was loaded allowed the baseline to load and then prevented configs from overriding.
```smalltalk
Metacello new
  baseline: 'Magritte3';
  lock.
```
@jbrichau maybe this can help you e.g. in a preloading script?
@dalehenrich?

**Comment by dalehenrich** [comment]
Yes `lock` is the preferred mechanism for declaring that a particular *version* of project should be the winner in all conflict situations ... so if locks work for you, then that is good ...

**Comment by seandenigris** [comment]
Cool. Can I do that *in* my baseline? I tried a global `#preLoadDoIt:`, but the hook doesn't seem to be triggered early enough…

**Comment by seandenigris** [comment]
Okay, so full workaround ATM is:
```smalltalk
BaselineOfMyCoolProject class>>#initialize
	Metacello new
		baseline: 'Magritte3';
		lock
```

**Comment by dalehenrich** [comment]
It's a bit spooky, but it certainly works as a workaround ... I still think SmalltalkCI should support locks/conflict resolution ... but it might be a bit of time before that support is implemented

**Comment by fniephaus** [comment]
@dalehenrich what'd be needed in smalltalkCI for this? I'm a little unfamiliar with the API for locking and conflict resolution.
For locking, it looks like the `SCIMetacelloLoadSpec` needs to be extended with a `#locking` dict that can contain a list of `#baselines` and `#configurations`? Maybe it also needs to be possible to set a version per baseline/configuration?
However, I have no idea what this might look like for conflict resolution, looks like actual code is needed?

**Comment by dalehenrich** [comment]
@fniephaus okay, the `onConflicts:` takes a Smalltalk block as an arg and the code in [@seandenigris' comment](https://github.com/hpi-swa/smalltalkCI/issues/333#issuecomment-375392861) is an illustration of a typical block.

I may have spoke too soon with regards to the `Metacello lock` ... on the one hand the `lock` is a special case of conflict resolution, where there is no need for conflict handling during a load - a lock is honored by default - so as in @seandenigris' case, a `lock` works in place of an `onConflict:` block - but I have a feeling that the fact that a lock works for @seandenigris does not imply that locks are needed or useful in general ...

The typical case for locks in my own development is when I'm have cloned a particular project and want all of the projects I load to use the clone ... and to be truthful, I would have to quiz @seandenigris as to how Magritte gets installed in his image before his "other project" is loaded, without following the dependencies of his "other project" ... I have not seen a transcript of the full load sequence, so I don't know what's going on during the load, so I can't guess how common it would be that a `lock` would be **preferable** over an `onConflict:` block ... My guess is that the fact that a `lock` *works* in this case is a lucky coincidence and that in the general case `onConflict:` will be more generally useful ...

**Comment by seandenigris** [comment]
In fact, I only explored `lock` because there seemed to be no way to use `onConflict:`. IIRC my project loads Magritte via baseline, and then a dependency loads the config version some layers down.

**Comment by dalehenrich** [comment]
... @seandenigris ... okay so I would agree that in your use case, `onConflict:` **is** the correct solution ... so I would claim that `lock` support is not necessary for SmalltalkCI, but `onConflict:` support would be useful for those special cases where (as @seandenigris' use case illustrates) an `onConflict:` is used to load the project itself ... and if I am not mistaken, @seandenigris is actually loading the **same** version of Magritte, but one project loads Magritte from github and another project loads Magritte from smalltalkhub ...

**Comment by estebanlm** [comment]
better later than never. 
there is a fix in the pipeline :)

**Comment by estebanlm** [comment]
in fact, I will need to send a new PR.
While this changes are ok, I need to also expose `ignoreImage` api and make it match also with `MetacelloIgnorePackageLoaded`.
Otherwise, packages coming from cypress or tonel will not be loaded anyway (since they will be already found in image).

**Comment by dalehenrich** [comment]
For metadataless filetree/cypress/tonel packages, I thought that if the **MetacelloCypressBaselineProject** class is used as the `projectClass` for a baseline that Metacello would unconditionally load the packages ... at least this is the preferred solution for the metadataless projects ... with that said, I think that #ignoreImage would work as well -- setting `projectClass` is not that well-known of a solution

**Comment by fniephaus** [comment]
Fixes with #399.
